### PR TITLE
ci: add share solo image cache in github workflow (new)

### LINKED ISSUE #4803: Use shared GitHub cache to warm Solo image cache in CI
## Problem

Many CI workflows repeatedly run `solo cache image pull`, downloading the same Solo dependency images from Hashgraph, GHCR, Docker Hub, Quay, and related registries. This adds latency to jobs and increases exposure to registry throttling or transient registry failures.

## Proposal

Add a shared GitHub Actions cache layer that stores Solo image cache artifacts and uses it as a buffer before falling back to registry pulls. Workflows should restore the cache early, run `solo cache image pull` only for cache misses or changed versions, then save/update the cache for later jobs.

## Goals

- Reduce elapsed time spent in `solo cache image pull` across CI workflows.
- Reduce repeated pulls from external registries for identical Solo/CN/MN/BN/relay/explorer versions.
- Keep cache keys version-aware so stale images are not reused across incompatible version changes.
- Provide a reusable workflow action or shared workflow snippet so all relevant workflows use the same cache behavior.

## Acceptance Criteria

- A shared cache key strategy covers Solo version, consensus node version, mirror node version, block node version, relay version, explorer version, and any registry mirror inputs that affect image selection.
- CI workflows restore the shared image cache before `solo cache image pull`.
- `solo cache image pull` can reuse restored artifacts and only pull missing images.
- Workflows save the warmed cache after successful image pull/cache population.
- Migration, one-shot, performance, examples, and nightly workflows use the common cache path instead of duplicating logic.
- Cache fallback remains safe: if GitHub cache is missing, stale, or corrupt, workflows still pull from registries and continue normally.

## Notes

This should be tracked separately from migration test fixes. The implementation can build on the `ci-share-solo-image-cache` branch/PR work if that branch is still available.

### PR DESCRIPTION
**Pull request by JeffreyDallas**
## Description

This pull request changes the following:

* Adds a reusable `.github/actions/solo-image-cache` composite action for restoring and saving Solo image archive caches with stable cache keys derived from Solo version files and `resources/config/solo-cache-images-target.yaml`.
* Wires Solo image cache restore/save steps into the main E2E, examples, migration, timing, performance, GCS, nightly, and Windows one-shot workflows, including the required GitHub Actions cache permissions.
* Extends Solo image cache target rendering to include chart-internal Solo images (`solo-cheetah` and `solo-containers`) and adds image-load verification so cache loads fail fast when expected images are not present in the kind cluster runtime.
* Fixes Podman-backed kind support for cache verification and archive loading, including rootful Podman clusters created through `sudo` in CI.
* Raises the one-shot REST Java Hikari pool limit in `resources/mirror-node-hikari-limits.yaml` to reduce Windows runner readiness failures under load.
* Adds unit coverage for Docker, Podman, and rootful Podman kind image listing/loading behavior.

### Related Issues

* Closes #4803
* Related to: N/A

### Pull request (PR) checklist

* [x] This PR added tests (unit, integration, and/or end-to-end)
* [ ] This PR updated documentation
* [x] This PR added no TODOs or commented out code
* [x] This PR has no breaking changes
* [x] Any technical debt has been documented as a separate issue and linked to this PR
* [x] Any `package.json` changes have been explained to and approved by a repository manager
* [x] All related issues have been linked to this PR
* [x] All changes in this PR are included in the description
* [x] When this PR merges the commits will be squashed and the title will be used as the commit message, the 'commit message guidelines' below have been followed

### Testing

* [x] This PR added unit tests
* [ ] This PR added integration/end-to-end tests
* [x] These changes required manual testing that is documented below
* [x] Anything not tested is documented

The following manual testing was done:

* `npx mocha 'test/unit/integration/container-engine/docker-client.test.ts' 'test/unit/commands/one-shot/orchestrator/deploy/deploy-argv-builders.test.ts'`
* `npx tsc --noEmit`
* `npx eslint src/integration/container-engine/docker-client.ts test/unit/integration/container-engine/docker-client.test.ts`
* `npx prettier --check resources/mirror-node-hikari-limits.yaml src/integration/container-engine/docker-client.ts test/unit/integration/container-engine/docker-client.test.ts`
* Investigated failing GitHub Actions runs for PR #4886 and used their logs to validate the Windows readiness and Podman/rootful kind cache-load fixes.

The following was not tested:

* Full local Windows/WSL one-shot E2E was not run locally; validation is expected through the GitHub Actions Windows workflow rerun.
* Full local repo-wide workflow matrix was not run locally; validation is expected through the PR checks.

<details>
<summary>
Commit message guidelines
</summary>
We use 'Conventional Commits' to ensure that our commit messages are easy to read, follow a consistent format, and for automated release note generation. Please follow the guidelines below when writing your commit messages:

1. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a breaking API change (correlating with MAJOR in Semantic Versioning). A BREAKING CHANGE can be part of commits of any type.  NOTE: currently breaking changes will only bump the MAJOR version.
2. The title is prefixed with one of the following:

| Prefix    | Description                                         | Semantic Version Update | Captured in Release Notes |
|-----------|-----------------------------------------------------|-------------------------|---------------------------|
| feat:     | a new feature                                       | MINOR                   | Yes                       |
| fix:      | a bug fix                                           | PATCH                   | Yes                       |
| perf:     | performance                                         | PATCH                   | Yes                       |
| refactor: | code change that isn't feature or fix               | none                    | No                        |
| test:     | adding missing tests                                | none                    | No                        |
| docs:     | changes to documentation                            | none                    | Yes                        |
| build:    | changes to build process                            | none                    | No                        |
| ci:       | changes to CI configuration                         | none                    | No                        |
| style:    | formatting, missing semi-colons, etc                | none                    | No                        |
| chore:    | updating grunt tasks etc; no production code change | none                    | No                        |

</details>

### CODE REVIEW

File: src/integration/container-engine/docker-client.ts [hunk 1]
```diff
@@ -16,10 +16,16 @@ import {LoadImageArchiveOptionsBuilder} from '../kind/model/load-image-archive/l
 import {LoadImageArchiveOptions} from '../kind/model/load-image-archive/load-image-archive-options.js';
 import {Architecture} from '../../business/utils/architecture.js';
 
+type ContainerEngineCommand = {
```
Comment (jan-milenkov) [inline src/integration/container-engine/docker-client.ts]: Move to separate file, this will also cause lint errors soon. as there is a PR being merged for it

Comment (JeffreyDallas) [reply src/integration/container-engine/docker-client.ts]: updated

File: version.ts [hunk 2]
```diff
@@ -45,6 +45,11 @@ export const REDIS_IMAGE_VERSION: string = constants.getEnvironmentVariable('RED
 export const REDIS_SENTINEL_IMAGE_VERSION: string =
   constants.getEnvironmentVariable('REDIS_SENTINEL_IMAGE_VERSION') || '8.2.2';
 
+// Image versions embedded in the solo-deployment Helm chart (SOLO_CHART_VERSION).
+// These must stay in sync with charts/solo-deployment/values.yaml whenever SOLO_CHART_VERSION bumps.
+export const SOLO_CHEETAH_VERSION: string = constants.getEnvironmentVariable('SOLO_CHEETAH_VERSION') || '0.4.5';
+export const SOLO_CONTAINERS_VERSION: string = constants.getEnvironmentVariable('SOLO_CONTAINERS_VERSION') || '0.46.0';
```
Comment (jeromy-cannon) [inline version.ts:51]: `SOLO_CHEETAH_VERSION` and `SOLO_CONTAINERS_VERSION` (lines 50-51) are both read via `getEnvironmentVariable()`. CLAUDE.md requires adding both to `docs/site/content/en/docs/env.md` with name, short description, and default value. Neither is documented there — please add both entries before merging.

File: .github/workflows/flow-pull-request-checks.yaml [hunk 3]
```diff
@@ -38,7 +38,7 @@ on:
 permissions:
   id-token: write
   contents: read
-  actions: read
+  actions: write
```
Comment (jeromy-cannon) [inline .github/workflows/flow-pull-request-checks.yaml:41]: This workflow only runs on `pull_request` events, and the save step in `zxc-e2e-test.yaml` is already gated by `github.event_name != 'pull_request'`, so no cache writes ever happen here. `actions: read` would satisfy least-privilege.
```suggestion
  actions: read
```

File: src/integration/container-engine/docker-client.ts [hunk 4]
```diff
@@ -95,4 +107,95 @@ export class DockerClient implements ContainerEngineClient {
...
+  }
+
+  private async loadImageArchiveIntoPodmanBackedCluster(
+    kindExecutable: string,
+    archivePath: string,
+    clusterName: string,
+    engineCommand: ContainerEngineCommand,
+  ): Promise<void> {
+    const kindArguments: string[] = ['load', 'image-archive', archivePath, '--name', clusterName];
+    const pathEnvironment: string = `${PathEx.dirname(kindExecutable)}${PathEx.delimiter}${process.env.PATH}`;
```
Comment (jeromy-cannon) [inline src/integration/container-engine/docker-client.ts]: `process.env.PATH` is typed `string | undefined` — the template literal produces `"…undefined"` if PATH is unset. Low real-world risk on Linux CI, but a `|| ''` guard keeps TypeScript strict-mode clean:
```suggestion
    const pathEnvironment: string = `${PathEx.dirname(kindExecutable)}${PathEx.delimiter}${process.env.PATH || ''}`;
```

File: .github/workflows/flow-standard-runner-one-shot-windows.yaml [hunk 5]
```diff
@@ -244,6 +245,14 @@ jobs:
           $env:PATH = "$env:USERPROFILE\.solo\bin;$env:PATH"
           npm run solo -- init --dev
 
+      - name: Restore Solo Image Cache
```
Comment (jeromy-cannon) [inline .github/workflows/flow-standard-runner-one-shot-windows.yaml]: The cache key embeds `RUNNER_OS=Windows` and `RUNNER_ARCH`, so no Linux-saved archive is consumable here. `zxc-e2e-test.yaml` gates saves to `inputs.runner != 'windows-latest'`, and this workflow has no save step — meaning the Windows cache key is never populated. Is this restore step expected to always miss? If Windows image caching is out of scope, it might be cleaner to skip the restore step rather than always take a cache miss.

File: resources/mirror-node-hikari-limits.yaml [hunk 6]
```diff
@@ -17,8 +17,8 @@ restjava:
     spring:
       datasource:
         hikari:
-          minimumIdle: 1
-          maximumPoolSize: 2
+          minimumIdle: 2
+          maximumPoolSize: 4
```
Comment (jeromy-cannon) [inline resources/mirror-node-hikari-limits.yaml]: Doubling the REST Java Hikari pool seems like it could be a workaround. What's the connection demand per Windows test run that justifies `maximumPoolSize: 4`? Should this be fixed upstream in the Mirror Node chart configuration instead?

File: .github/workflows/flow-pull-request-checks.yaml (hunk 3 above)
Comment (JeffreyDallas) [reply .github/workflows/flow-pull-request-checks.yaml:41]: it need to call zxc e2e test workflow, will fail if no write permission
https://github.com/hiero-ledger/solo/actions/runs/28414536516

File: src/integration/container-engine/docker-client.ts (hunk 4 above)
Comment (JeffreyDallas) [reply src/integration/container-engine/docker-client.ts]: updated

File: .github/workflows/flow-standard-runner-one-shot-windows.yaml (hunk 5 above)
Comment (JeffreyDallas) [reply .github/workflows/flow-standard-runner-one-shot-windows.yaml]: removed

File: version.ts (hunk 2 above)
Comment (JeffreyDallas) [reply version.ts:51]: no such file `docs/site/content/en/docs/env.md` exists

File: resources/mirror-node-hikari-limits.yaml (hunk 6 above)
Comment (JeffreyDallas) [reply resources/mirror-node-hikari-limits.yaml]: reverted the change

File: src/integration/container-engine/docker-client.ts
Comment (jan-milenkov) [inline src/integration/container-engine/docker-client.ts:1]: this is combining logic for podman and docker inside the `DockerClient`

Comment (JeffreyDallas) [reply src/integration/container-engine/docker-client.ts:1]: updated

File: version.ts (hunk 2 above)
Comment (jeromy-cannon) [reply version.ts:51]: true, need to tweak my claude skill, but does solo-docs need to be updated appropriately and can we link the solo-docs PR if so?

Comment (JeffreyDallas) [reply version.ts:51]: PR link: https://github.com/hiero-ledger/solo-docs/pull/249

**Review by jeromy-cannon — APPROVED** [review]
